api/api-internal-environments-v1/src/main/java/com/thoughtworks/go/apiv1/internalenvironments/InternalEnvironmentsControllerV1.java

```diff
@@ -108,17 +109,15 @@ public String indexMergedEnvironments(Request request, Response response) throws
     String updateAgentAssociation(Request request, Response response) {
         String envName = request.params("env_name");
         EnvironmentConfig envConfig = environmentConfigService.getEnvironmentConfig(envName);
-        List<String> uuids = getAgentUuids(request);
-        agentService.updateAgentsAssociationOfEnvironment(envConfig, uuids);
-        return renderMessage(response, 200, "Environment '" + envName + "' updated successfully!");
-    }
-
-    private List<String> getAgentUuids(Request request) {
         JsonReader jsonReader = GsonTransformer.getInstance().jsonReaderFrom(request.body());
-        Optional<List<String>> uuids = jsonReader.readStringArrayIfPresent("uuids");
-        if (uuids.isPresent()) {
-            return uuids.get();
+        JsonReader agents = jsonReader.readJsonObject("agents");
+
+        List<String> uuidsToAssociate = agents.readStringArrayIfPresent("add").orElse(Collections.emptyList());
+        List<String> uuidsToRemove = agents.readStringArrayIfPresent("remove").orElse(Collections.emptyList());
+
+        if (!uuidsToAssociate.isEmpty() || !uuidsToRemove.isEmpty()) {
+            agentService.updateAgentsAssociationOfEnvironment(envConfig, uuidsToAssociate, uuidsToRemove);
         }
-        throw halt(422, MessageJson.create(format("Json `%s` does not contain property '%s'.", new Gson().fromJson(request.body(), JsonElement.class), "uuids")));
+        return renderMessage(response, 200, "Environment '" + envName + "' updated successfully!");
```

Comment:
How about `return renderMessage(response, 200, EntityType.Environment("bob").updateSuccessful());` ?

We will need to define `updateSuccessful` utility method similar to `deleteSuccessful`